larcorealg/Geometry/OpDetGeo.cxx

```diff
@@ -24,23 +29,17 @@ namespace geo{
 
   //-----------------------------------------
   OpDetGeo::OpDetGeo(TGeoNode const& node, geo::TransformationMatrix&& trans)
-    : fTrans(std::move(trans))
-  {
-    fOpDetNode = &node;
-
-    fCenter = toWorldCoords(geo::origin<LocalPoint_t>());
-  }
-
+    : fTrans{ makeTrans(std::move(trans)) }
+    , fOpDetNode{ &node }
+    , fShape{ makeShape(*fOpDetNode) }
```

Comment:
Not to squelch your enthusiasm for C++17 😉, but mixing `std::variant`-based polymorphism with virtual-based polymorphism introduces more points of maintenance than it removes.  I think what is required is that the `fSpecs` object can be created/overwritten during the `UpdateAfterSorting` call.  Instead of using the `std::variant`, is it possible to bind to the correct specs-maker function in the constructor and then just call it where necessary?  For example:

```cpp
fSpecsMaker{ specsMakerFor(*fOpDetNodex) } // Creates specs-maker bounded to pointer of derived shape type
// ... and then in UpdateAfterSorting
fSpecs = fSpecsMaker(...);
```
I've not thought this through fully, so you may end up with something that's more complicated than what you have.

Reply:
I was indeed experimenting with `variant`...
There is a lot to say here.

First, I agree that mixing this `variant` "polymorphism" with ROOT's is awkward, and I would rather have only one of them used. I did not go all the way through, and at the conclusion of this discussion I will gladly change the class to adopt only one of them.

Then: the "traditional" polymorphism via inheritance has at this level gone stale: all the uses except one (in `geo::OpDetGeo::PrintOpDetInfo()`) are via `dynamic_cast`ing to the final type. Given the nature of `geo::OpDetGeo`, which shape is used is determined at construction and never changed: all these `dynamic_cast()` just rediscover over and over the same answer, and if anything make the code a tiny bit more obscure.
For this reason, a more sensible choice in my opinion is to store which one of the shapes is present, and then use that static information when needed. The `variant` object has exactly this information and a bit of interface around it, so it felt like a reasonable option rather than a less expressive integer or a more expressive but still less immediate enumerator.

Moreover, it has been my dark master plan (never carried to the end) since my first rehaul of `geo::OpdetGeo` (not this one: years ago) to detach from the underlying polymorphism and to attempt to give `geo::OpDetGeo()` a uniform interface. This may also be biasing my choice against dynamic casts. In my ideal world, there would be no access to `Shape()` or `GetNode()`. Admittedly those ones are still there and the whole thing may be a lost fight, since the performance of photodetectors depends _a lot_ on the details of their shape.

Last, I am not very fond of having a data member in the class that is used exactly once in the object lifetime, at ~construction~ initialisation. I think your approach can straightforwardly be moved to the `UpdateAfterSorting()` though, to initialise a local "spec maker" to be used and then disposed of.

Ok, after this maximally boring minimal essay, the real point of discussion is the second one: which approach is the most appropriate.
As you can guess, I lean toward the more static one, with `dynamic_cast()` confined in a small little corner as an unavoidable evil.

Reply:
(I moved the pull request to draft under the persuasion that any outcome of this discussion will require either a new implementation or some finalisation work before reaching merge quality)

Reply:
Note that I still need feedback on this point.

Reply:
Gianluca, I've been thinking about this for awhile, and I'm having trouble coming up with an elegant solution.  Let's go with what you have.